Parsimony/Parser.cs

```diff
@@ -0,0 +1,212 @@
+        {
+            var tokens = (input ?? throw new ArgumentNullException(nameof(input))).ToList();
+
+            var options = new T();
+            var arguments = new List<string>(tokens.Count);
+            var result = new ParseResult<T>(options, arguments);
+            return Parse(result, tokens);
+        }
+
+        /// <summary>
+        /// Parses a <typeparamref name="T"/> from <paramref name="input"/>.
+        /// </summary>
+        /// <param name="input">The tokens to parse.</param>
+        /// <returns>The <see cref="ParseResult{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="input"/> is <c>null</c>.
+        /// </exception>
+        public ParseResult<T> Parse(params string[] input)
+        {
+            return Parse(input as IEnumerable<string>);
+        }
+
+        private ParseResult<T> Parse(ParseResult<T> result, List<string> input)
+        {
+            var state = new ParserState(result, input);
+            while (state.Input.Count > 0)
+                state = Parse(state);
+            return state.Result;
+        }
+
+        private ParserState Parse(ParserState state)
+        {
+            return ConsumeShortNameOption(state)
+                ?? ConsumeLongNameOption(state)
+                ?? ConsumeArgument(state)
+                ?? state;
+        }
+
+        private ParserState? ConsumeShortNameOption(ParserState state)
+        {
+            var options = state.Result.Options;
+            var arguments = state.Result.Arguments.ToList();
+            var input = state.Input;
+
+            if (input.Count == 0)
+                return null;
+
+            var match = _shortNameOptionPatern.Match(input[0]);
+            if (!match.Success)
+                return null;
+
+            var optionName = match.Groups["option"].Value[0];
+            var option = _options.FirstOrDefault(o => o.Spec.ShortName == optionName);
+            if (option == null)
+                // TODO: Custom exception
+                // TODO: Support for unknown options as arguments
+                throw new Exception($"Unknown argument '-{optionName}'");
+
+            // Consume the token
+            input = input.Skip(1).ToList();
+
+            var optionType = option.Spec.OptionType;
+            var typedOption = typeof(OptionSpec<,>).MakeGenericType(new[] { typeof(T), optionType });
+            var set = typedOption.GetMethod("Set", new[] { typeof(T), optionType });
+
+            if (optionType == typeof(bool))
+            {
+                set.Invoke(option.Spec, new object[] { options, true });
+                option.Set = true;
+
+                if (match.Groups["rest"].Success)
+                    input.Insert(1, $"-{match.Groups["rest"].Value}");
+
+                return new ParserState(new ParseResult<T>(options, arguments), input);
+            }
+
+            string nextToken()
```

Comment:
I preferred the closure over `input` instead treating it as an out param. I suppose the modified value for input could be returned in a `{ string Token, IEnumerable<string> Input }` type though

Reply:
Yeah I was thinking that you'd just return input.Skip(1).ToList() since you don't use the results of RemoveAt() and it's the same number of copies regardless since it doesn't just bump the head ref

Reply:
But yeah never out variables